File: mobile/lib/models/survey.dart
import 'package:hive/hive.dart';

import 'survey_photo.dart';

part 'survey.g.dart';

class SurveyStatus {
  static const draft = 'DRAFT';
  static const pending = 'PENDING';
  static const uploading = 'UPLOADING';
  static const synced = 'SYNCED';
  static const failed = 'FAILED';
}

@HiveType(typeId: 0)
class Survey extends HiveObject {
  @HiveField(0)
  String clientId; // UUID v4

  @HiveField(1)
  String vehicleNumber;

  @HiveField(2)
  String notes;

  @HiveField(3)
  double? latitude;

  @HiveField(4)
  double? longitude;

  @HiveField(5)
  String status; // DRAFT | PENDING | UPLOADING | SYNCED | FAILED

  @HiveField(6)
  List<SurveyPhoto> photos;

  @HiveField(7)
  DateTime createdAt;

  @HiveField(8)
  DateTime updatedAt;

  @HiveField(9)
  String? remoteId; // server _id once synced

  @HiveField(10)
  int uploadedPhotoCount;

  @HiveField(11)
  String? lastError;

  Survey({
    required this.clientId,
    required this.vehicleNumber,
    required this.notes,
    required this.latitude,
    required this.longitude,
    required this.status,
    required this.photos,
    required this.createdAt,
    required this.updatedAt,
    this.remoteId,
    this.uploadedPhotoCount = 0,
    this.lastError,
  });

  int get totalPhotos => photos.length;

  double get uploadProgress {
    if (photos.isEmpty) return 0;
    return uploadedPhotoCount / photos.length;
  }
}

File: backend/src/models/Survey.js
'use strict';

const mongoose = require('mongoose');

const STATUS = Object.freeze({
  DRAFT: 'DRAFT',
  PENDING: 'PENDING',
  UPLOADING: 'UPLOADING',
  SYNCED: 'SYNCED',
  FAILED: 'FAILED',
});

const PointSchema = new mongoose.Schema(
  {
    type: { type: String, enum: ['Point'], default: 'Point' },
    coordinates: {
      type: [Number],
      validate: {
        validator: (v) =>
          Array.isArray(v) &&
          v.length === 2 &&
          v[0] >= -180 && v[0] <= 180 &&
          v[1] >= -90 && v[1] <= 90,
        message: 'coordinates must be [longitude, latitude] within valid bounds',
      },
    },
  },
  { _id: false },
);

const SurveySchema = new mongoose.Schema(
  {
    vehicleNumber: { type: String, required: true, uppercase: true, trim: true, index: true },
    surveyorId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
    },
    surveyorName: { type: String, required: true },
    notes: { type: String, default: '' },
    location: { type: PointSchema, default: undefined },
    status: {
      type: String,
      enum: Object.values(STATUS),
      default: STATUS.PENDING,
      index: true,
    },
    photoCount: { type: Number, default: 0 },
    clientId: { type: String, index: true, sparse: true },
    syncedAt: { type: Date, default: null },
  },
  { timestamps: true },
);

SurveySchema.index({ vehicleNumber: 1, createdAt: -1 });
SurveySchema.index({ surveyorId: 1, createdAt: -1 });
SurveySchema.index({ status: 1, createdAt: -1 });
SurveySchema.index({ location: '2dsphere' });
SurveySchema.index(
  { surveyorId: 1, clientId: 1 },
  { unique: true, partialFilterExpression: { clientId: { $exists: true, $ne: null } } },
);

const Survey = mongoose.model('Survey', SurveySchema);

module.exports = Survey;
module.exports.STATUS = STATUS;
